guides/hack/01-getting-started/03-starting-a-real-project.md

```diff
@@ -0,0 +1,332 @@
+
+A good starting point is to:
+- [install Composer]
+- create an empty `.hhconfig` file
+- create `src/` and `tests/` subdirectories
+- configure autoloading
+- use Composer to install the common dependencies and tools:
+
+### Autoloading
+
+In HHVM, there is no 'build' step as such; each file is processed as needed.
+Currently, HHVM needs to be given a map of what files define which classes,
+functions and so on - for example, to execute the code `new Foo()`, HHVM needs
+to know that the class `Foo` is defined in `src/Foo.hack`.
+
+[hhvm-autoload] generates this map, using an `hh_autoload.json` configuration
+file. For most projects, a minimal example is:
+
+```JSON
+{
+  "roots": [
+    "src/"
+  ],
+  "devRoots": [
+    "tests/"
+  ]
+}
+```
+
+Once this configuration file is created, `vendor/bin/hh-autoload` can be executed
+to generate or update the map, which is created as `vendor/autoload.hack`
+
+### An Example
+
+```
+$ touch .hhconfig
+$ mkdir bin src tests
+$ cat > hh_autoload.json
+{
+  "roots": [
+    "src/"
+  ],
+  "devRoots": [
+    "tests/"
+  ]
+}
+$ composer require hhvm/hsl hhvm/hhvm-autoload
+Using version ^4.0 for hhvm/hsl
+Using version ^2.0 for hhvm/hhvm-autoload
+./composer.json has been created
+Loading composer repositories with package information
+Updating dependencies (including require-dev)
+Package operations: 2 installs, 0 updates, 0 removals
+  - Installing hhvm/hsl (v4.0.0): Loading from cache
+  - Installing hhvm/hhvm-autoload (v2.0.3): Loading from cache
+Writing lock file
+Generating autoload files
+/var/folders/3l/2yk1tgkn7xdd76bs547d9j90fcbt87/T/tmp.xaQwE1xE/vendor/autoload.hack
+$ composer require --dev hhvm/hhast hhvm/hacktest facebook/fbexpect
+Using version ^4.0 for hhvm/hhast
+Using version ^1.4 for hhvm/hacktest
+Using version ^2.5 for facebook/fbexpect
+./composer.json has been updated
+Loading composer repositories with package information
+Updating dependencies (including require-dev)
+Package operations: 7 installs, 0 updates, 0 removals
+  - Installing facebook/difflib (v1.1): Loading from cache
+  - Installing hhvm/hsl-experimental (v4.0.1): Loading from cache
+  - Installing hhvm/type-assert (v3.3.1): Loading from cache
+  - Installing facebook/hh-clilib (v2.1.0): Loading from cache
+  - Installing hhvm/hhast (v4.0.4): Loading from cache
+  - Installing hhvm/hacktest (v1.4): Loading from cache
+  - Installing facebook/fbexpect (v2.5.1): Loading from cache
+Writing lock file
+Generating autoload files
+/private/var/folders/3l/2yk1tgkn7xdd76bs547d9j90fcbt87/T/tmp.xaQwE1xE/vendor/autoload.hack
+$
+```
+
+### Adding Functions or Classes
+
+As a toy example, we're going to create a function that squares a vector of
+numbers; save the following as `src/square_vec.hack`:
+
+```Hack
+use namespace HH\Lib\Vec;
+
+function square_vec<T as num>(vec<T> $numbers): vec<int> {
+  return Vec\map($numbers, $number ==> $number * $number);
+}
```

Comment:
Hack with generics and type constraints seems a little intense for newbies. Could we use concrete types?

```php
function square_vec(vec<float> $numbers): vec<int> {
  return Vec\map($numbers, $number ==> $number * $number);
}
```